qiskit/ml/algorithms/quantum_kernel.py

```diff
@@ -0,0 +1,272 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Quantum Kernel Algorithm"""
+
+from typing import Optional, Union
+
+import logging
+
+import numpy as np
+
+from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
+from qiskit.circuit import ParameterVector
+from qiskit.providers import Backend, BaseBackend
+
+from qiskit.aqua import QuantumInstance
+
+logger = logging.getLogger(__name__)
+
+
+class QuantumKernel:
+    """
+    Quantum Kernel
+
+    The general task of machine learning is to find and study patterns in data. For many
+    algorithms, the datapoints are better understood in a higher dimensional feature space,
+    through the use of a kernel function:
+
+    K(x, y) = <f(x), f(y)>.
+
+    Here K is the kernel function, x, y are n dimensional inputs. f is a map from n-dimension
+    to m-dimension space. <x,y> denotes the dot product. Usually m is much larger than n.
+
+    The quantum kernel algorithm calculates a kernel matrix, given datapoints x and y and feature
+    map f, all of n dimension. This kernel matrix can then be used in classical machine learning
+    algorithms such as support vector classification, spectral clustering or ridge regression.
+    """
+
+    BATCH_SIZE = 1000
+
+    def __init__(self,
+                 feature_map: QuantumCircuit,
+                 enforce_psd: bool = False,
+                 quantum_instance: Optional[
+                     Union[QuantumInstance, BaseBackend, Backend]] = None) -> None:
+        """
+        Args
+            feature_map:       # parametrised circuit to be used as the feature map
+            enforce_psd:       # project to closest positive semidefinite matrix if x = y
+            quantum_instance:  # Quantum Instance or Backend
+        """
+        # super().__init__(quantum_instance)
+
+        self._feature_map = feature_map
+        self._enforce_psd = enforce_psd
+        self._quantum_instance = quantum_instance
+
+    @property
+    def feature_map(self) -> QuantumCircuit:
+        """ Returns feature map """
+        return self._feature_map
+
+    @feature_map.setter
+    def feature_map(self, feature_map: QuantumCircuit):
+        """ Sets feature map """
+        self._feature_map = feature_map
+
+    @property
+    def quantum_instance(self) -> QuantumInstance:
+        """ Returns quantum instance """
+        return self._quantum_instance
+
+    @quantum_instance.setter
+    def quantum_instance(self, quantum_instance: Union[Backend,
+                                                       BaseBackend, QuantumInstance]) -> None:
+        """ Sets quantum instance """
+        if isinstance(quantum_instance, (BaseBackend, Backend)):
+            self._quantum_instance = QuantumInstance(quantum_instance)
+        else:
+            self._quantum_instance = quantum_instance
+
+    def _construct_circuit(self, x, y, is_statevector_sim=False):
```

Comment:
Added `measurement` flag and `measurement_basis` arguments in 3878862. `statevector_sim` flag is different though, because if we use `statevector_simulator`, we only build circuits for  Psi(x1)|0>, rather than rather than Psi(x2)^dagger Psi(x1)|0>.